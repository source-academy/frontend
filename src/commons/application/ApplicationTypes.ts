import { Chapter, Language, SourceError, Variant } from 'js-slang/dist/types';

import { AcademyState } from '../../features/academy/AcademyTypes';
import { AchievementState } from '../../features/achievement/AchievementTypes';
import { DashboardState } from '../../features/dashboard/DashboardTypes';
import { Grading } from '../../features/grading/GradingTypes';
import { PlaygroundState } from '../../features/playground/PlaygroundTypes';
import { PlaybackStatus, RecordingStatus } from '../../features/sourceRecorder/SourceRecorderTypes';
import { WORKSPACE_BASE_PATHS } from '../../pages/fileSystem/createInBrowserFileSystem';
import { Assessment } from '../assessment/AssessmentTypes';
import { FileSystemState } from '../fileSystem/FileSystemTypes';
import Constants from '../utils/Constants';
import { createContext } from '../utils/JsSlangHelper';
import {
  DebuggerContext,
  WorkspaceLocation,
  WorkspaceManagerState,
  WorkspaceState
} from '../workspace/WorkspaceTypes';
import { ExternalLibraryName } from './types/ExternalTypes';
import { SessionState } from './types/SessionTypes';

export type OverallState = {
  readonly academy: AcademyState;
  readonly achievement: AchievementState;
  readonly application: ApplicationState;
  readonly playground: PlaygroundState;
  readonly session: SessionState;
  readonly workspaces: WorkspaceManagerState;
  readonly dashboard: DashboardState;
  readonly fileSystem: FileSystemState;
};

export type ApplicationState = {
  readonly environment: ApplicationEnvironment;
};

export type Story = {
  story: string;
  playStory: boolean;
};

export type GameState = {
  collectibles: { [id: string]: string };
  completed_quests: string[];
};

/**
 * An output while the program is still being run in the interpreter. As a
 * result, there are no return values or SourceErrors yet. However, there could
 * have been calls to display (console.log) that need to be printed out.
 */
export type RunningOutput = {
  type: 'running';
  consoleLogs: string[];
};

/**
 * An output which reflects the program which the user had entered. Not a true
 * Output from the interpreter, but simply there to let he user know what had
 * been entered.
 */
export type CodeOutput = {
  type: 'code';
  value: string;
};

/**
 * An output which represents a program being run successfully, i.e. with a
 * return value at the end. A program can have either a return value, or errors,
 * but not both.
 */
export type ResultOutput = {
  type: 'result';
  value: any;
  consoleLogs: string[];
  runtime?: number;
  isProgram?: boolean;
};

/**
 * An output which represents a program being run unsuccessfully, i.e. with
 * errors at the end. A program can have either a return value, or errors, but
 * not both.
 */
export type ErrorOutput = {
  type: 'errors';
  errors: SourceError[];
  consoleLogs: string[];
};

export type InterpreterOutput = RunningOutput | CodeOutput | ResultOutput | ErrorOutput;

export enum ApplicationEnvironment {
  Development = 'development',
  Production = 'production',
  Test = 'test'
}

export enum Role {
  Student = 'student',
  Staff = 'staff',
  Admin = 'admin'
}

/**
 * Defines the languages available for use on Source Academy,
 * including Source sublanguages and other languages e.g. full JS.
 * For external libraries, see ExternalLibrariesTypes.ts
 */
export interface SALanguage extends Language {
  displayName: string;
}

const variantDisplay: Map<Variant, string> = new Map([
  [Variant.TYPED, 'Typed'],
  [Variant.WASM, 'WebAssembly'],
  [Variant.NON_DET, 'Non-Det'],
  [Variant.CONCURRENT, 'Concurrent'],
  [Variant.LAZY, 'Lazy'],
  [Variant.GPU, 'GPU'],
  [Variant.NATIVE, 'Native'],
  [Variant.EXPLICIT_CONTROL, 'Explicit-Control']
]);

export const fullJSLanguage: SALanguage = {
  chapter: Chapter.FULL_JS,
  variant: Variant.DEFAULT,
  displayName: 'full JavaScript'
};

export const fullTSLanguage: SALanguage = {
  chapter: Chapter.FULL_TS,
  variant: Variant.DEFAULT,
  displayName: 'full TypeScript'
};

export const htmlLanguage: SALanguage = {
  chapter: Chapter.HTML,
  variant: Variant.DEFAULT,
  displayName: 'HTML'
};

export const styliseSublanguage = (chapter: Chapter, variant: Variant = Variant.DEFAULT) => {
  switch (chapter) {
    case Chapter.FULL_JS:
      return fullJSLanguage.displayName;
    case Chapter.FULL_TS:
      return fullTSLanguage.displayName;
    case Chapter.HTML:
      return htmlLanguage.displayName;
    default:
      return `Source \xa7${chapter}${
        variantDisplay.has(variant) ? ` ${variantDisplay.get(variant)}` : ''
      }`;
  }
};

export const sublanguages: Language[] = [
  { chapter: Chapter.SOURCE_1, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_1, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_1, variant: Variant.WASM },
  { chapter: Chapter.SOURCE_1, variant: Variant.LAZY },
  { chapter: Chapter.SOURCE_1, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_2, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_2, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_2, variant: Variant.LAZY },
  { chapter: Chapter.SOURCE_2, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_3, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_3, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_3, variant: Variant.CONCURRENT },
  { chapter: Chapter.SOURCE_3, variant: Variant.NON_DET },
  { chapter: Chapter.SOURCE_3, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_4, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_4, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_4, variant: Variant.GPU },
  { chapter: Chapter.SOURCE_4, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_4, variant: Variant.EXPLICIT_CONTROL }
];

export const sourceLanguages: SALanguage[] = sublanguages.map(sublang => {
  return {
    ...sublang,
    displayName: styliseSublanguage(sublang.chapter, sublang.variant)
  };
});

export const defaultLanguages = sourceLanguages.filter(
  sublang => sublang.variant === Variant.DEFAULT
);

export const variantLanguages = sourceLanguages.filter(
  sublang => sublang.variant !== Variant.DEFAULT
);

export const isSourceLanguage = (chapter: Chapter) =>
  [Chapter.SOURCE_1, Chapter.SOURCE_2, Chapter.SOURCE_3, Chapter.SOURCE_4].includes(chapter);

const currentEnvironment = (): ApplicationEnvironment => {
  switch (process.env.NODE_ENV) {
    case 'development':
      return ApplicationEnvironment.Development;
    case 'production':
      return ApplicationEnvironment.Production;
    default:
      return ApplicationEnvironment.Test;
  }
};

export const defaultAcademy: AcademyState = {
  gameCanvas: undefined
};

export const defaultApplication: ApplicationState = {
  environment: currentEnvironment()
};

export const defaultDashboard: DashboardState = {
  gradingSummary: {
    cols: [],
    rows: []
  }
};

export const defaultAchievement: AchievementState = {
  achievements: [],
  goals: [],
  users: [],
  assessmentOverviews: []
};

export const defaultPlayground: PlaygroundState = {
  githubSaveInfo: { repoName: '', filePath: '' }
};

export const defaultEditorValue = '// Type your program in here!';

/**
 * Create a default IWorkspaceState for 'resetting' a workspace.
 * Takes in parameters to set the js-slang library and chapter.
 *
 * @param workspaceLocation the location of the workspace, used for context
 */
export const createDefaultWorkspace = (workspaceLocation: WorkspaceLocation): WorkspaceState => ({
  autogradingResults: [],
  context: createContext<WorkspaceLocation>(
    Constants.defaultSourceChapter,
    [],
    workspaceLocation,
    Constants.defaultSourceVariant
  ),
  isFolderModeEnabled: false,
  activeEditorTabIndex: 0,
  editorTabs: [
    {
      filePath: ['playground', 'sicp'].includes(workspaceLocation)
        ? getDefaultFilePath(workspaceLocation)
        : undefined,
      value: ['playground', 'sourcecast', 'githubAssessments'].includes(workspaceLocation)
        ? defaultEditorValue
        : '',
      highlightedLines: [],
      breakpoints: []
    }
  ],
  programPrependValue: '',
  programPostpendValue: '',
  editorSessionId: '',
  isEditorReadonly: false,
  editorTestcases: [],
  externalLibrary: ExternalLibraryName.NONE,
  execTime: 1000,
  output: [],
  replHistory: {
    browseIndex: null,
    records: [],
    originalValue: ''
  },
  replValue: '',
  sharedbConnected: false,
  stepLimit: 1000,
  globals: [],
  isEditorAutorun: false,
  isRunning: false,
  isDebugging: false,
  enableDebugging: true,
  debuggerContext: {} as DebuggerContext
});

const defaultFileName = 'program.js';
export const getDefaultFilePath = (workspaceLocation: WorkspaceLocation) =>
  `${WORKSPACE_BASE_PATHS[workspaceLocation]}/${defaultFileName}`;

export const defaultWorkspaceManager: WorkspaceManagerState = {
  assessment: {
    ...createDefaultWorkspace('assessment'),
    currentAssessment: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  grading: {
    ...createDefaultWorkspace('grading'),
    currentSubmission: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  playground: {
    ...createDefaultWorkspace('playground'),
    usingSubst: false,
    activeEditorTabIndex: 0,
    editorTabs: [
      {
        filePath: getDefaultFilePath('playground'),
        value: defaultEditorValue,
        highlightedLines: [],
        breakpoints: []
      }
    ]
  },
  sourcecast: {
    ...createDefaultWorkspace('sourcecast'),
    audioUrl: '',
    codeDeltasToApply: null,
    currentPlayerTime: 0,
    description: null,
    inputToApply: null,
    playbackData: {
      init: {
        editorValue: '',
        chapter: Chapter.SOURCE_1,
        externalLibrary: ExternalLibraryName.NONE
      },
      inputs: []
    },
    playbackDuration: 0,
    playbackStatus: PlaybackStatus.paused,
    sourcecastIndex: null,
    title: null,
    uid: null
  },
  sourcereel: {
    ...createDefaultWorkspace('sourcereel'),
    playbackData: {
      init: {
        editorValue: '',
        chapter: Chapter.SOURCE_1,
        externalLibrary: ExternalLibraryName.NONE
      },
      inputs: []
    },
    recordingStatus: RecordingStatus.notStarted,
    timeElapsedBeforePause: 0,
    timeResumed: 0
  },
  sicp: {
    ...createDefaultWorkspace('sicp'),
    usingSubst: false,
    activeEditorTabIndex: 0,
    editorTabs: [
      {
        filePath: getDefaultFilePath('sicp'),
        value: defaultEditorValue,
        highlightedLines: [],
        breakpoints: []
      }
    ]
  },
  githubAssessment: {
    ...createDefaultWorkspace('githubAssessment'),
    hasUnsavedChanges: false
  }
};

export const defaultSession: SessionState = {
  courses: [],
  group: null,
  gameState: {
    completed_quests: [],
    collectibles: {}
  },
  xp: 0,
  allUserXp: undefined,
  story: {
    story: '',
    playStory: false
  },
  assessments: new Map<number, Assessment>(),
  assessmentOverviews: undefined,
  agreedToResearch: undefined,
  sessionId: Date.now(),
  githubOctokitObject: { octokit: undefined },
  gradingOverviews: undefined,
  gradings: new Map<number, Grading>(),
  notifications: []
};

export const defaultFileSystem: FileSystemState = {
  inBrowserFileSystem: null
};

export const defaultState: OverallState = {
  academy: defaultAcademy,
  achievement: defaultAchievement,
  application: defaultApplication,
  dashboard: defaultDashboard,
  playground: defaultPlayground,
  session: defaultSession,
  workspaces: defaultWorkspaceManager,
  fileSystem: defaultFileSystem
};
