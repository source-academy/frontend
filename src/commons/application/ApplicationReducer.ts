import { Action, Reducer } from 'redux';

import { ApplicationState, defaultApplication } from './ApplicationTypes';

export const ApplicationReducer: Reducer<ApplicationState> = (
  state = defaultApplication,
  action: Action
) => {
  return state;
};
