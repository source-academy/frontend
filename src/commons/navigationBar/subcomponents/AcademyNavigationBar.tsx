import { Alignment, Classes, Icon, Navbar, NavbarGroup } from '@blueprintjs/core';
import { IconName, IconNames } from '@blueprintjs/icons';
import classNames from 'classnames';
import * as React from 'react';
import { NavLink } from 'react-router-dom';
import { AssessmentType } from 'src/commons/assessment/AssessmentTypes';
import { assessmentTypeLink } from 'src/commons/utils/ParamParseHelper';

import { Role } from '../../application/ApplicationTypes';
import NotificationBadgeContainer from '../../notificationBadge/NotificationBadgeContainer';
import { filterNotificationsByType } from '../../notificationBadge/NotificationBadgeHelper';

type OwnProps = {
  courseId: number;
  role: Role;
  assessmentTypes?: AssessmentType[];
};

const AcademyNavigationBar: React.FunctionComponent<OwnProps> = props => (
  <Navbar className="NavigationBar secondary-navbar">
    <NavbarGroup align={Alignment.LEFT}>
      {props.assessmentTypes?.map((assessmentType, idx) => (
        <NavLink
          to={`/courses/${props.courseId}/${assessmentTypeLink(assessmentType)}`}
          activeClassName={Classes.ACTIVE}
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
          key={idx}
        >
          <Icon icon={icons[idx]} />
          <div className="navbar-button-text hidden-xs hidden-sm">{assessmentType}</div>
          <NotificationBadgeContainer
            notificationFilter={filterNotificationsByType(assessmentType)}
            disableHover={true}
          />
        </NavLink>
      ))}
    </NavbarGroup>
    {props.role === Role.Admin || props.role === Role.Staff ? (
      <NavbarGroup align={Alignment.RIGHT}>
        <NavLink
          to={`/courses/${props.courseId}/groundcontrol`}
          activeClassName={Classes.ACTIVE}
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
        >
          <Icon icon={IconNames.SATELLITE} />
          <div className="navbar-button-text hidden-xs hidden-sm">Ground Control</div>
        </NavLink>

        <NavLink
          to={`/courses/${props.courseId}/dashboard`}
          activeClassName={Classes.ACTIVE}
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
        >
          <Icon icon={IconNames.GLOBE} />
          <div className="navbar-button-text hidden-xs hidden-sm">Dashboard</div>
        </NavLink>

        <NavLink
          to={`/courses/${props.courseId}/sourcereel`}
          activeClassName={Classes.ACTIVE}
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
        >
          <Icon icon={IconNames.MOBILE_VIDEO} />
          <div className="navbar-button-text hidden-xs hidden-sm hidden-md">Sourcereel</div>
        </NavLink>

        {props.role === Role.Admin && (
          <NavLink
            to={`/courses/${props.courseId}/xpcalculation`}
            activeClassName={Classes.ACTIVE}
            className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
          >
            <Icon icon={IconNames.CALCULATOR} />
            <div className="navbar-button-text hidden-xs hidden-sm">XP Calculation</div>
          </NavLink>
        )}

        <NavLink
          to={`/courses/${props.courseId}/grading`}
          activeClassName={Classes.ACTIVE}
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
        >
          <Icon icon={IconNames.ENDORSED} />
          <div className="navbar-button-text hidden-xs hidden-sm hidden-md">Grading</div>
          <NotificationBadgeContainer
            notificationFilter={filterNotificationsByType('Grading')}
            disableHover={true}
          />
        </NavLink>

        <NavLink
          to={`/courses/${props.courseId}/storysimulator`}
          activeClassName={Classes.ACTIVE}
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
        >
          <Icon icon={IconNames.CROWN} />
          <div className="navbar-button-text hidden-xs hidden-sm hidden-md">Story Simulator</div>
        </NavLink>

        {props.role === Role.Admin && (
          <NavLink
            to={`/courses/${props.courseId}/adminpanel`}
            activeClassName={Classes.ACTIVE}
            className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
          >
            <Icon icon={IconNames.SETTINGS} />
            <div className="navbar-button-text hidden-xs hidden-sm hidden-md hidden-lg">
              Admin Panel
            </div>
          </NavLink>
        )}
      </NavbarGroup>
    ) : null}
  </Navbar>
);

export const icons: IconName[] = [
  IconNames.FLAME,
  IconNames.LIGHTBULB,
  IconNames.PREDICTIVE_ANALYSIS,
  IconNames.COMPARISON,
  IconNames.MANUAL,
  IconNames.GRAPH,
  IconNames.LAB_TEST,
  IconNames.CALCULATOR
];

export default AcademyNavigationBar;
