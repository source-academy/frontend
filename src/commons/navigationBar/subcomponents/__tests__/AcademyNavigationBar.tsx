import { shallow } from 'enzyme';

import { Role } from '../../../application/ApplicationTypes';
import AcademyNavigationBar from '../AcademyNavigationBar';

test('MissionControl, GroundControl, Sourcereel, StorySimulator, Dashboard, Grading and AdminPanel NavLinks do NOT render for Role.Student', () => {
  const props = {
    role: Role.Student,
    notifications: [],
    assessmentTypes: ['Missions', 'Quests', 'Paths', 'Contests', 'Others'],
    courseId: 0
  };
  const tree = shallow(<AcademyNavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/grading' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/groundcontrol' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/sourcereel' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/storysimulator' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/dashboard' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/adminpanel' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/xpcalculation' }).exists())
  ).toHaveLength(0);
});

test('MissionControl, GroundControl, Sourcereel, StorySimulator, Dashboard and Grading NavLinks render for Role.Staff', () => {
  const props = {
    role: Role.Staff,
    notifications: [],
    assessmentTypes: ['Missions', 'Quests', 'Paths', 'Contests', 'Others'],
    courseId: 0
  };
  const tree = shallow(<AcademyNavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/grading' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/groundcontrol' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/sourcereel' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/storysimulator' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/dashboard' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/xpcalculation' }).exists())
  ).toHaveLength(0);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/adminpanel' }).exists())
  ).toHaveLength(0);
});

test('MissionControl, GroundControl, Sourcereel, StorySimulator, Dashboard, Grading, XP Calculation and AdminPanel NavLinks render for Role.Admin', () => {
  const props = {
    role: Role.Admin,
    notifications: [],
    assessmentTypes: ['Missions', 'Quests', 'Paths', 'Contests', 'Others'],
    courseId: 0
  };
  const tree = shallow(<AcademyNavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/grading' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/groundcontrol' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/sourcereel' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/storysimulator' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/dashboard' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/xpcalculation' }).exists())
  ).toHaveLength(1);
  expect(
    tree.filterWhere(shallowTree => shallowTree.find({ to: '/courses/0/adminpanel' }).exists())
  ).toHaveLength(1);
});
