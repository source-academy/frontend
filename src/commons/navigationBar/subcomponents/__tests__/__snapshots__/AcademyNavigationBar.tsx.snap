// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionControl, GroundControl, Sourcereel, StorySimulator, Dashboard and Grading NavLinks render for Role.Staff 1`] = `
<Blueprint4.Navbar
  className="NavigationBar secondary-navbar"
>
  <Blueprint4.NavbarGroup
    align="left"
  >
    <NavLink
      className={[Function]}
      title="Missions"
      to="/courses/0/missions"
    >
      <Blueprint4.Icon
        icon="flame"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Missions
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Quests"
      to="/courses/0/quests"
    >
      <Blueprint4.Icon
        icon="lightbulb"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Quests
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Paths"
      to="/courses/0/paths"
    >
      <Blueprint4.Icon
        icon="predictive-analysis"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Paths
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Contests"
      to="/courses/0/contests"
    >
      <Blueprint4.Icon
        icon="comparison"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Contests
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Others"
      to="/courses/0/others"
    >
      <Blueprint4.Icon
        icon="manual"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Others
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
  </Blueprint4.NavbarGroup>
  <Blueprint4.NavbarGroup
    align="right"
  >
    <NavLink
      className={[Function]}
      title="Ground Control"
      to="/courses/0/groundcontrol"
    >
      <Blueprint4.Icon
        icon="satellite"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Ground Control
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Dashboard"
      to="/courses/0/dashboard"
    >
      <Blueprint4.Icon
        icon="globe"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Dashboard
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Sourcereel"
      to="/courses/0/sourcereel"
    >
      <Blueprint4.Icon
        icon="mobile-video"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md"
      >
        Sourcereel
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Grading"
      to="/courses/0/grading"
    >
      <Blueprint4.Icon
        icon="endorsed"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md"
      >
        Grading
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Story Simulator"
      to="/courses/0/storysimulator"
    >
      <Blueprint4.Icon
        icon="crown"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md"
      >
        Story Simulator
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Notifications"
      to="/courses/0/notipreference"
    >
      <Blueprint4.Icon
        icon="notifications"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md hidden-lg"
      >
        Notifications
      </div>
    </NavLink>
  </Blueprint4.NavbarGroup>
</Blueprint4.Navbar>
`;

exports[`MissionControl, GroundControl, Sourcereel, StorySimulator, Dashboard, Grading and AdminPanel NavLinks do NOT render for Role.Student 1`] = `
<Blueprint4.Navbar
  className="NavigationBar secondary-navbar"
>
  <Blueprint4.NavbarGroup
    align="left"
  >
    <NavLink
      className={[Function]}
      title="Missions"
      to="/courses/0/missions"
    >
      <Blueprint4.Icon
        icon="flame"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Missions
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Quests"
      to="/courses/0/quests"
    >
      <Blueprint4.Icon
        icon="lightbulb"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Quests
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Paths"
      to="/courses/0/paths"
    >
      <Blueprint4.Icon
        icon="predictive-analysis"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Paths
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Contests"
      to="/courses/0/contests"
    >
      <Blueprint4.Icon
        icon="comparison"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Contests
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Others"
      to="/courses/0/others"
    >
      <Blueprint4.Icon
        icon="manual"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Others
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
  </Blueprint4.NavbarGroup>
  <Blueprint4.NavbarGroup
    align="right"
  >
    <NavLink
      className={[Function]}
      title="Notifications"
      to="/courses/0/notipreference"
    >
      <Blueprint4.Icon
        icon="notifications"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md hidden-lg"
      >
        Notifications
      </div>
    </NavLink>
  </Blueprint4.NavbarGroup>
</Blueprint4.Navbar>
`;

exports[`MissionControl, GroundControl, Sourcereel, StorySimulator, Dashboard, Grading, XP Calculation and AdminPanel NavLinks render for Role.Admin 1`] = `
<Blueprint4.Navbar
  className="NavigationBar secondary-navbar"
>
  <Blueprint4.NavbarGroup
    align="left"
  >
    <NavLink
      className={[Function]}
      title="Missions"
      to="/courses/0/missions"
    >
      <Blueprint4.Icon
        icon="flame"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Missions
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Quests"
      to="/courses/0/quests"
    >
      <Blueprint4.Icon
        icon="lightbulb"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Quests
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Paths"
      to="/courses/0/paths"
    >
      <Blueprint4.Icon
        icon="predictive-analysis"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Paths
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Contests"
      to="/courses/0/contests"
    >
      <Blueprint4.Icon
        icon="comparison"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Contests
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Others"
      to="/courses/0/others"
    >
      <Blueprint4.Icon
        icon="manual"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Others
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
  </Blueprint4.NavbarGroup>
  <Blueprint4.NavbarGroup
    align="right"
  >
    <NavLink
      className={[Function]}
      title="Ground Control"
      to="/courses/0/groundcontrol"
    >
      <Blueprint4.Icon
        icon="satellite"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Ground Control
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Dashboard"
      to="/courses/0/dashboard"
    >
      <Blueprint4.Icon
        icon="globe"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm"
      >
        Dashboard
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Sourcereel"
      to="/courses/0/sourcereel"
    >
      <Blueprint4.Icon
        icon="mobile-video"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md"
      >
        Sourcereel
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Grading"
      to="/courses/0/grading"
    >
      <Blueprint4.Icon
        icon="endorsed"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md"
      >
        Grading
      </div>
      <NotificationBadge
        disableHover={true}
        notificationFilter={[Function]}
      />
    </NavLink>
    <NavLink
      className={[Function]}
      title="Story Simulator"
      to="/courses/0/storysimulator"
    >
      <Blueprint4.Icon
        icon="crown"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md"
      >
        Story Simulator
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="XP Calculation"
      to="/courses/0/xpcalculation"
    >
      <Blueprint4.Icon
        icon="calculator"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md hidden-lg"
      >
        XP Calculation
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Admin Panel"
      to="/courses/0/adminpanel"
    >
      <Blueprint4.Icon
        icon="settings"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md hidden-lg"
      >
        Admin Panel
      </div>
    </NavLink>
    <NavLink
      className={[Function]}
      title="Notifications"
      to="/courses/0/notipreference"
    >
      <Blueprint4.Icon
        icon="notifications"
      />
      <div
        className="navbar-button-text hidden-xs hidden-sm hidden-md hidden-lg"
      >
        Notifications
      </div>
    </NavLink>
  </Blueprint4.NavbarGroup>
</Blueprint4.Navbar>
`;
